packages/core/src/model.ts

```diff
@@ -80,23 +110,122 @@ export class ChartModel {
 	 */
 	setData(newData) {
 		const sanitizedData = this.sanitize(Tools.clone(newData));
-		const dataLabels = this.generateDataLabels(sanitizedData);
+		const dataGroups = this.generateDataGroups(sanitizedData);
 
 		this.set({
 			data: sanitizedData,
-			dataLabels
+			dataGroups
 		});
 
 		return sanitizedData;
 	}
 
-	generateDataLabels(newData) {
-		const dataLabels = {};
-		newData.datasets.forEach(dataset => {
-			dataLabels[dataset.label] = Configuration.legend.items.status.ACTIVE;
+	/*
+	 * Data groups
+	*/
+	protected updateAllDataGroups() {
+		// If allDataGroups hasn't been initialized yet
+		// Set it to the current set of data groups
+		if (!this.allDataGroups) {
+			this.allDataGroups = this.getDataGroups().map(group => group.name);
+		} else {
+			// Loop through current data groups
+			this.getDataGroups().forEach(dataGroup => {
+				// If group name hasn't been stored yet, store it
+				if (this.allDataGroups.indexOf(dataGroup.name) === -1) {
+					this.allDataGroups.push(dataGroup.name);
+				}
+			});
+		}
+	}
+
+	protected generateDataGroups(data) {
+		const { groupIdentifier } = this.getOptions().data;
+		const { ACTIVE } = Configuration.legend.items.status;
+
+		const uniqueDataGroups = map(data, datum => datum[groupIdentifier]).keys();
+		return uniqueDataGroups.map(groupName => ({
+			name: groupName,
+			status: ACTIVE
+		}));
+	}
+
+	getDataGroups() {
+		return this.get("dataGroups");
+	}
+
+
+	getDataGroupNames() {
+		return this.get("dataGroups").map(dataGroup => dataGroup.name);
+	}
+
+	getGroupedData() {
+		const displayData = this.getDisplayData();
+		const groupedData = {};
+		const { groupIdentifier } = this.getOptions().data;
+
+		displayData.map(datum => {
+			const group = datum[groupIdentifier];
+			if (groupedData[group] !== null && groupedData[group] !== undefined) {
+				groupedData[group].push(datum);
+			} else {
+				groupedData[group] = [datum];
+			}
 		});
 
-		return dataLabels;
+		return Object.keys(groupedData)
+			.map(groupName => {
+				return {
+					name: groupName,
+					data: groupedData[groupName]
+				};
+			});
+	}
+
+	getDataValuesGroupedByKeys() {
+		const options = this.getOptions();
+		const { groupIdentifier } = options.data;
+
+		const displayData = this.getDisplayData();
+		const domainIdentifier = this.services.cartesianScales.getDomainIdentifier();
+		const rangeIdentifier = this.services.cartesianScales.getRangeIdentifier();
+
+		const stackKeys = map(displayData, datum => datum[domainIdentifier]).keys();
+		const dataGroupNames = this.getDataGroupNames();
+
+		return stackKeys.map(key => {
+			const correspondingValues = { sharedStackKey: key };
+			dataGroupNames.forEach(dataGroupName => {
+				const correspondingDatum = displayData.find(datum => {
+					return datum[groupIdentifier] === dataGroupName &&
+						`${datum[domainIdentifier]}` === key;
```

Comment:
Looks like
```suggestion
						datum[domainIdentifier] === key;
```
should work too?

Or perhaps
```suggestion
						datum[domainIdentifier] as string === key;
```

Reply:
that's done to make sure that everything is stringified. e.g. if you had a `number` or a `date` type

Reply:
@zvonimirfras the logic would actually break if we get rid of that for time-series charts

Reply:
I'd maybe suggest a more explicit `.toString()` to make it clear we need it to be a string.